### DISCUSSION TITLE
Streamlink Filmon.py

### DISCUSSION THREAD

**Discussion by Saule2006** [streamlink/streamlink#4360] [Q&A] [answered]
Does anybody has the working solution for Filmon with Streamlink? I am getting this "error: ('No cipher can be selected.',)"

**Comment by bastimeyer** [comment, marked as answer]
> error: ('No cipher can be selected.',)

Post the **full debug log** by setting [`--loglevel=debug`](https://streamlink.github.io/cli.html#cmdoption-loglevel), and don't remove any lines.

The plugin was just fixed last week ago:
https://github.com/streamlink/streamlink/pull/4335

There hasn't been a new release since then, so you'll have to use the recent development version.
https://streamlink.github.io/install.html

**Comment by Saule2006** [comment]
root@ubuntub:/# streamlink "https://www.filmon.com/tv/snooker-legends" best --player-external-http --player-external-http-port 2050 --loglevel=debug
[cli][info] streamlink is running as root! Be careful!
[cli][debug] OS:         Linux-4.4.0-190-generic-x86_64-with-Ubuntu-16.04-xenial
[cli][debug] Python:     3.7.9
[cli][debug] Streamlink: 3.1.1
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.filmon.com/tv/snooker-legends
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player-external-http=True
[cli][debug]  --player-external-http-port=2050
[cli][info] Found matching plugin filmon for URL https://www.filmon.com/tv/snooker-legends
error: ('No cipher can be selected.',)

**Reply by bastimeyer** [reply]
> Streamlink: 3.1.1

As said, there hasn't been a new release since then, so you'll have to use the recent development version.

```
$ streamlink -l debug 'https://www.filmon.com/tv/snooker-legends' best
[cli][debug] OS:         Linux-5.16.10-1-git-x86_64-with-glibc2.35
[cli][debug] Python:     3.10.2
[cli][debug] Streamlink: 3.1.1+12.gaad49e8
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.filmon.com/tv/snooker-legends
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin filmon for URL https://www.filmon.com/tv/snooker-legends
[plugins.filmon][debug] Found cached channel ID: 3325
[cli][info] Available streams: low (worst), high (best)
[cli][info] Opening stream: high (hls-filmon)
[cli][info] Starting player: mpv
[plugins.filmon][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[plugins.filmon][debug] Reloading FilmOn channel playlist: 3325
[stream.hls][debug] First Sequence: 226657; Last Sequence: 226662
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 226660; End Sequence: None
[stream.hls][debug] Adding segment 226660 to queue
[stream.hls][debug] Adding segment 226661 to queue
[stream.hls][debug] Adding segment 226662 to queue
[stream.hls][debug] Segment 226660 complete
[cli.output][debug] Opening subprocess: mpv --force-media-title=https://www.filmon.com/tv/snooker-legends -
[stream.hls][debug] Segment 226661 complete
[stream.hls][debug] Segment 226662 complete
[cli][debug] Writing stream to output
```

**Comment by Saule2006** [comment]
Can you post full working filmon.py ?

**Reply by bastimeyer** [reply]
The new plugin implementation relies on other new changes. You can find everything in the linked pull request. Good luck.

**Comment by Saule2006** [comment]
[cli][debug] OS:         Linux-4.4.0-190-generic-x86_64-with-Ubuntu-16.04-xenial
[cli][debug] Python:     3.7.9
[cli][debug] Streamlink: 3.1.1+14.g19fdc30.dirty
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.filmon.com/tv/snooker-legends
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player-external-http=True
[cli][debug]  --player-external-http-port=2050
[cli][debug]  --http-no-ssl-verify=True
[cli][info] Found matching plugin filmon for URL https://www.filmon.com/tv/snooker-legends
error: ('No cipher can be selected.',)

**Reply by bastimeyer** [reply]
As you can read in https://github.com/streamlink/streamlink/pull/4335 where the plugin got updated, filmon.com is using an outdated TLS version and on top of that encryption ciphers which are considered insecure by today's standards. The plugin therefore now forces Python to accept these insecure ciphers by setting the list of ciphers to `DEFAULT:@SECLEVEL=1`. These insecure ciphers are disabled by default on Python 3.10, so this is a necessary step. Otherwise, this depends on your system's OpenSSL configuration, but as said, this is not the case anymore on Python 3.10.

Since you are using Ubuntu Xenial 16.04, I had a look at which OpenSSL version is running on this old distro and it turns out that OpenSSL 1.0.2 doesn't support the `@SECLEVEL=n` cipher suite aliases and it's only available since 1.1.1:
- https://www.openssl.org/docs/man1.1.1/man1/ciphers.html
- https://www.openssl.org/docs/man1.0.2/man1/ciphers.html

This means that if we want to support older OpenSSL versions, we'll have to change the ciphers list to include explicit ciphers (by type) instead of cipher suites by security level.
https://en.wikipedia.org/wiki/OpenSSL#Major_version_releases

**Comment by Saule2006** [comment]
You are right - openssl was 1.0.x
I have upgraded it, but this won't help:
root@ubuntub:~/.local/bin# openssl version
OpenSSL 1.1.1c  28 May 2019


[cli][info] streamlink is running as root! Be careful!
[cli][debug] OS:         Linux-4.4.0-190-generic-x86_64-with-Ubuntu-16.04-xenial
[cli][debug] Python:     3.7.9
[cli][debug] Streamlink: 3.1.1+14.g19fdc30.dirty
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.filmon.com/tv/snooker-legends
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player-external-http=True
[cli][debug]  --player-external-http-port=2050
[cli][info] Found matching plugin filmon for URL https://www.filmon.com/tv/snooker-legends
error: ('No cipher can be selected.',)

**Reply by bastimeyer** [reply]
Python 3.7 is not available on Ubuntu 16.04 without adding external package repositories. Depending on where you've got the python package from and how it was built, it is not guaranteed that it's using your system's OpenSSL version, so upgrading it won't do anything. It is very likely that it is not using your system's OpenSSL version, because otherwise that would mean that the python 3.7 package would set it as a package dependency, and it and any other package on your system would rely on a different OpenSSL version that is shipped via the custom package repo you're (probably) using, which would be bad. So far, you have not posted any relevant informations about your system's environment.

You can check whether OpenSSL is dynamically linked or not by running
```
$ ldd -v /usr/lib/python3.7/lib-dynload/_ssl.cpython-37-x86_64-linux-gnu.so
```
(or whatever the lib's file name is on your system) and checking whether it includes links to OpenSSL via `libssl.so.X.Y`.

Ubuntu 16.04 is not actively supported anymore, and the Python 3.7 version you're using is outdated as well, so you should upgrade your system if you want any kind of support.
https://endoflife.date/ubuntu
https://docs.python.org/release/3.7.12/whatsnew/changelog.html#python-3-7-9-final

Apart from the cipher suite names issue with OpenSSL 1.0.2 that I've mentioned above, there is nothing that can be done here.

**Comment by Saule2006** [comment]
Does anybody has sound?

**Reply by mkbloke** [reply]
Yes. The plugin is working fine.

**Comment by Saule2006** [comment]
Very strange. The sound is not working. Video is fine

[cli][debug] OS:         Linux-5.4.0-104-generic-x86_64-with-glibc2.29
[cli][debug] Python:     3.8.10
[cli][debug] Streamlink: 3.1.1+14.g19fdc30
[cli][debug] Requests(2.27.1), Socks(1.7.1), Websocket(1.2.3)
[cli][debug] Arguments:
[cli][debug]  url=https://www.france.tv/france-5/direct.html
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player-external-http=True
[cli][debug]  --player-external-http-port=3001
[cli][info] Found matching plugin pluzz for URL https://www.france.tv/france-5/direct.html
[plugins.pluzz][debug] Country: FR
[plugins.pluzz][debug] Video ID: 45007886-f3ff-4b3e-9706-1ef1014c5a60
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: 144p (worst), 216p, 360p, 540p, 720p (best)
[cli][info] Starting server, access with one of:
[cli][info]  http://127.0.0.1:3001/
[cli][info]  http://127.0.1.1:3001/
[cli][info] Got HTTP request from VLC/3.0.12 LibVLC/3.0.12
[cli][info] Opening stream: 720p (hls)
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] First Sequence: 28379003; Last Sequence: 28380802
[stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 28380800; End Sequence: None
[stream.hls][debug] Adding segment 28380800 to queue
[stream.hls][debug] Adding segment 28380801 to queue
[stream.hls][debug] Adding segment 28380802 to queue
[stream.hls][debug] Segment 28380800 complete
[cli][debug] Writing stream to player
[stream.hls][debug] Segment 28380801 complete
[stream.hls][debug] Segment 28380802 complete
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Segments in this playlist are encrypted
[stream.hls][debug] Adding segment 28380803 to queue
[stream.hls][debug] Segment 28380803 complete

**Reply by bastimeyer** [reply]
You aren't even using the filmon plugin which this thread is about. Stop abusing the discussion forum like that or I'll have to lock it, because this is completely unnecessary and spammy noise.

Now to your issue with the pluzz plugin (france.tv). This plugin requires muxing of separate video and audio stream sources, and FFmpeg is a requirement for that. If you don't have FFmpeg installed and available on the `PATH` env var, this won't work without setting the `--ffmpeg-ffmpeg=/path/to/ffmpeg` CLI argument, and it can only output the video stream. What needs to be improved here is that this should be logged by the MuxedHLSStream implementation.

**Reply by Saule2006** [reply]
I am very sorry. You are right. This is a new ubuntu installation an I have forget to install the ffmpeg. Thank You.

**Comment by Saule2006** [comment]
What does this line mean?
[stream.hls][debug] Segments in this playlist are encrypted
